internal/output/result.go

```diff
@@ -0,0 +1,126 @@
+package output
+
+import (
+	"encoding/json"
+
+	"github.com/google/osv-scanner/pkg/models"
+	"golang.org/x/exp/slices"
+)
+
+type pkgWithSource struct {
+	Package models.PackageInfo
+	Source  models.SourceInfo
+}
+
+// Custom implementation of this unique set map to allow it to serialize to JSON
+type pkgSourceSet map[pkgWithSource]struct{}
+
+func (pss *pkgSourceSet) MarshalJSON() ([]byte, error) {
+	res := []pkgWithSource{}
+
+	for v := range *pss {
+		res = append(res, v)
+	}
+
+	return json.Marshal(res)
+}
+
+func (pss *pkgSourceSet) UnmarshalJSON(data []byte) error {
+	aux := []pkgWithSource{}
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+
+	*pss = make(pkgSourceSet)
+	for _, pws := range aux {
+		(*pss)[pws] = struct{}{}
+	}
+
+	return nil
+}
+
+type groupedVuln struct {
+	DisplayID    string
+	PkgSource    pkgSourceSet
+	AliasedVulns map[string]models.Vulnerability
+}
+
+// groupFixedVersions builds the fixed versions for each ID Group, with keys formatted like so:
+// `Source:ID`
+func groupFixedVersions(flattened []models.VulnerabilityFlattened) map[string][]string {
+	groupFixedVersions := map[string][]string{}
+
+	// Get the fixed versions indexed by each group of vulnerabilities
+	// Prepend source path as same vulnerability in two projects should be counted twice
+	// Remember to sort and compact before displaying later
+	for _, vf := range flattened {
+		groupIdx := vf.Source.String() + ":" + vf.GroupInfo.IndexString()
+		pkg := models.Package{
+			Ecosystem: models.Ecosystem(vf.Package.Ecosystem),
+			Name:      vf.Package.Name,
+		}
+		groupFixedVersions[groupIdx] =
+			append(groupFixedVersions[groupIdx], vf.Vulnerability.FixedVersions()[pkg]...)
+	}
+
+	// Remove duplicates
+	for k := range groupFixedVersions {
+		fixedVersions := groupFixedVersions[k]
+		slices.Sort(fixedVersions)
+		groupFixedVersions[k] = slices.Compact(fixedVersions)
+	}
+
+	return groupFixedVersions
+}
+
+// groupByVulnGroups creates a map over all vulnerability IDs, with aliased vuln IDs
+// pointing to the same groupVulns object
+func groupByVulnGroups(vulns *models.VulnerabilityResults) map[string]*groupedVuln {
```

Comment:
nit: this function name on its own is pretty confusing with two instances of "group". 

Perhaps rename this as `mapVulnIDsToSARIFFinding` or something similar?